Require variable names to be globally unique (#402)
unicode-org/message-format-wg: spec/formatting.md

To resolve the value of a _variable_,
its _name_ is used to identify either a local variable,
or a variable defined elsewhere.

> It is an error for the right-hand side of a local variable declaration to

Reviewer: the right and left hand sides have names in the ABNF. You should use those. In this case it should be something like:

```suggestion
The `expression` part of a `declaration` MUST NOT refer to a `variable` that is declared after it in the `message`.
```

(Technically we could just say "an `expression` MUST NOT refer to a `variable` that has not been assigned in a previous `declaration` or passed in...")

Reviewer 2: +1 to Addison's phrasing.

Author: I made an attempt to rephrase this. 

This would be much easier to specify if, in the grammar, `declaration` and `operand` were changed:

```
declaration = let s variable-definition [s] "=" [s] expression
operand = literal / variable-reference
option = name [s] "=" [s] operand
variable-definition = variable
variable-reference = variable
```

If this change were made, then I'd be able to write something like:

"The _variable_ of any `variable-reference` must be preceded by a full _declaration_ whose `variable-definition` has the same _variable_."

and could avoid referring to the enclosing context of "the operand".

refer to a local variable that's declared after it in the message.

> Variable names are required to be globally unique. That is,

Reviewer: Use normative language.

```suggestion
Each `variable` MUST be unique within a message. It is a (what type?) error if a `declaration` uses a name assigned by a previous `declaration`.
```

Reviewer 2: in the spirit of the generic comment I left on this PR, I suggest we use "constants" everywhere in this PR (and `const` keyword). (of course not for the passed-in ones)

Reviewer: @mihnita Note that I'm using production names from the ABNF. If we decided to make the `variables` constant then we should change the name of the production at the same time.

Author: Rewrote this; for now I'll defer the `const` vs. `let` question and try to focus on making the language precise.

for any local variable declaration of the form `let $v = e`:
* It is an error if `v` is the left-hand side of a local variable declaration
that appears before `let $v = e` in the message.

> * It is also an error if `v` is an externally defined variable.

Reviewer: We discussed this briefly in the recent call (talking about changing the sigil for external values). I think this is bad because it is easy for code to suddenly produce an error when the input data (which is dynamic) collides with a declaration.

Reviewer 2: Nitpick: `let $v = e` => `let $v = {$e}` (everywhere)

Author: @aphillips I agree that it would be preferable to change the sigil for externally-defined variables. If that change lands before this PR does (i.e. if #403 is resolved), then I'll update this PR to reflect that.

@mihnita I think my rewrites that I just pushed eliminate metavariables anyway, but if I end up re-adding any I'll keep that in mind.

    > when * {The value is not one.}
    > ```

  - **Variable Redefinition errors** occur when

>     the left-hand side of a variable declaration is already defined,

Reviewer: Get rid of "left-hand side". The formal name `variable` and the term `definition` work fine.

```suggestion
a `declaration` contains a `variable` that is already defined
```

    either by a previous local variable declaration,
    or by an external definition.

    > For example, attempting to format the following message

>     > must result in a Variable Redefinition error,

Reviewer: must => MUST

... although examples are by definition non-normative, which means you should avoid normative language.... so...

must result => results

    > because there is another declaration for `var`
    > that textually precedes
    > the declaration of `var` to be `{|horse|}`.
    >
    > ```
    > let $var = {|cart|}
    > let $var = {|horse|}
    > {The value is {$var}.}
    > ```
    >
    > Attempting to format the following message in a context
    > where `var` is externally defined

>     > should also result in a Variable Redefinition error.

Reviewer: see above. And "should" is super wrong 😉 

Just say "results" or "also results"

    >
    > ```
    > let $var = {|horse|}
    > {The value is {$var}.}
    > ```

>   - **Unknown Function errors** occur when an _expression_ includes

Reviewer 2: Nitpick: would be good to have a code example with a function here

Author: Do you mean a function in the declaration of `$var`, or a function in the expression that uses `$var`, or both?

> ```
> The value is {|horse|}.
> ```

When there are multiple _declarations_ for the same _variable_,
the fallback string is formatted based on the _expression_ of

> the first declaration of that _variable_.

Reviewer: what about if `var` were externally defined? And why is this a good fallback string?

Reviewer 2: > When there are multiple _declarations_ for the same _variable_,

That would be an error.

Author: > what about if `var` were externally defined?

Rewrote to include that case as well.

> And why is this a good fallback string?

See https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md?plain=1#L208

I wanted to be as consistent as possible with the existing spec, which says that fallback values should be formatted based on the RHS of a declaration. I thought the most consistent thing if there are multiple declarations is to pick one of them. Do you have another suggestion?

Author: > > When there are multiple _declarations_ for the same _variable_,
> 
> That would be an error.
> 

@mihnita I'm not sure what you mean -- doesn't there have to be a fallback string to replace any expression that resolves to an error?

> For example, attempting to format the following message:
>
> ```
> let $var = {|cart|}

> > let $var = {|horse|}

Reviewer 2: The second declaration (`let $var = {|horse|}`) would be an error.

Author: @mihnita I can't find the context for this comment since I've since pushed more commits -- I'm "resolving", but if the issue is still there, please comment again.
